### PR TITLE
many: implement new REST API: GET /2.0/interfaces

### PR DESCRIPTION
**Pull request by zyga**
This branch implements the new REST API for listing interface connections.

The /2.0/interfaces endpoint now returns an object with two lists, one for plugs and one for slots. This change allows clients to discover disconnected slots. Each of the plugs and slots also contains a list of connections (possibly empty). Each connection is described by a plug or slot "reference", which holds only two identifier fields, "snap" and either "plug" or "slot".

There are supporting changes in the backend, client API and client commands (snap interfaces).

### CODE REVIEW

File: client/interfaces.go [hunk 1]
```diff
@@ -26,30 +26,44 @@ import (
 
 // Plug represents a capacity offered by a snap.
 type Plug struct {
+	Snap string `json:"snap"`
 	// NOTE: the discrepancy between go and json field name is intentional.
```
Comment (niemeyer) [inline client/interfaces.go]: Please drop the note.

File: client/interfaces.go [hunk 2]
```diff
@@ -26,30 +26,44 @@ import (
...
+// PlugRef is a reference to a plug.
+type PlugRef struct {
+	Snap string `json:"snap"`
+	Name string `json:"plug"`
 }
 
 // Slot represents the potential of a given snap to connect to a given plug.
 type Slot struct {
+	Snap string `json:"snap"`
 	// NOTE: the discrepancy between go and json field name is intentional.
```
Comment (niemeyer) [inline client/interfaces.go]: Please drop the note.

File: client/interfaces.go [hunk 3]
```diff
@@ -26,30 +26,44 @@ import (
...
+	Snap string `json:"snap"`
+	Name string `json:"slot"`
 }
 
-// PlugConnections represents a single plug and slots that are connected to it.
-type PlugConnections struct {
-	Plug
-	Connections []Slot `json:"connections"`
+// InterfaceConnections contains information about all plugs, slots and their connections
+type InterfaceConnections struct {
```
Comment (niemeyer) [inline client/interfaces.go]: s/InterfaceConnections/Interfaces/

This lists plugs and slots, whether connected or not.

File: client/interfaces.go [hunk 4]
```diff
@@ -59,8 +73,8 @@ type InterfaceAction struct {
 	Slot   *Slot  `json:"slot,omitempty"`
 }
 
-// AllPlugs returns information about all the plugs and their connections.
-func (client *Client) AllPlugs() (connections []PlugConnections, err error) {
+// InterfaceConnections returns information about all the plugs, slots and their connections.
+func (client *Client) InterfaceConnections() (connections InterfaceConnections, err error) {
```
Comment (niemeyer) [inline client/interfaces.go]: s/InterfaceConnections/Interfaces/

Comment (niemeyer) [inline client/interfaces.go]: Also, s/connections/interfaces/

File: daemon/api.go [hunk 5]
```diff
@@ -919,41 +919,9 @@ func appIconGet(c *Command, r *http.Request) Response {
...
-		plugs = append(plugs, plugInfo{
-			Snap:        plug.Snap,
-			Name:        plug.Plug,
-			Interface:   plug.Interface,
-			Label:       plug.Label,
-			Connections: slots,
-		})
-	}
-	return SyncResponse(plugs)
+// getInterfaces returns a response with a list of all the plugs and slots and their connections.
```
Comment (niemeyer) [inline daemon/api.go]: `"returns a response with a list of all the plugs and slots"` => `"returns all plugs and slots"`

File: client/interfaces.go [hunk 6]
```diff
@@ -59,8 +73,8 @@ type InterfaceAction struct {
 	Slot   *Slot  `json:"slot,omitempty"`
 }
 
-// AllPlugs returns information about all the plugs and their connections.
-func (client *Client) AllPlugs() (connections []PlugConnections, err error) {
+// InterfaceConnections returns information about all the plugs, slots and their connections.
```
Comment (niemeyer) [inline client/interfaces.go]: `// Interfaces returns all plugs, slots and their connections.`

File: cmd/snap/cmd_interfaces.go [hunk 7]
```diff
@@ -59,34 +59,61 @@ func init() {
...
+			if x.Positionals.Query.Name != "" && x.Positionals.Query.Name != plug.Name {
 				continue
 			}
 			if x.Interface != "" && plug.Interface != x.Interface {
 				continue
 			}
-			fmt.Fprintf(w, "%s:%s\t", plug.Snap, plug.Plug.Name)
+			// The OS snap (always ubuntu-core) is special and enable abbreviated
+			// display syntax on the plug-side of the connection.
+			if plug.Snap != "ubuntu-core" {
```
Comment (niemeyer) [inline cmd/snap/cmd_interfaces.go]: Nitpick suggestion for coding style: in most cases it's easier for humans to read branching logic on the positive side.

So, first I'd list the "snap _is_ ubuntu-core" one, then the alternative.

File: interfaces/repo.go [hunk 8]
```diff
@@ -449,28 +449,73 @@ func (r *Repository) PlugConnections(snapName, plugName string) []*Slot {
...
+	var result []*Plug
+	for plug := range r.slotPlugs[slot] {
+		result = append(result, plug)
+	}
+	sort.Sort(byPlugSnapAndName(result))
+	return result
 }
 
-type bySlotSnapAndName []*Slot
+// InterfaceConnections returns object holding a lists of all the plugs and slots and their connections.
```
Comment (niemeyer) [inline interfaces/repo.go]: `// Interfaces returns all plugs, slots, and their connections.`

I wasn't considering having that and SlotRef/PlugRef in the internal APIs, but maybe that's a good idea.

File: interfaces/core.go [hunk 9]
```diff
@@ -27,22 +27,42 @@ import (
...
+	Connections []PlugRef              `json:"connections,omitempty"`
+}
+
+// SlotRef is a reference to a slot.
+type SlotRef struct {
+	Snap string `json:"snap"`
+	Name string `json:"slot"`
+}
+
+// InterfaceConnections contains information about all plugs, slots and their connections
```
Comment (niemeyer) [inline interfaces/core.go]: `// Interfaces holds information about a list of plugs and slots, and their connections.`

File: interfaces/repo.go [hunk 10]
```diff
@@ -449,28 +449,73 @@ func (r *Repository) PlugConnections(snapName, plugName string) []*Slot {
 	return result
 }
 
-// Support for sort.Interface
-
-type byPlugSnapAndName []*Plug
+// SlotConnections returns all of the plugs that are connected a given slot.
+func (r *Repository) SlotConnections(snapName, slotName string) []*Plug {
```
Comment (niemeyer) [inline interfaces/repo.go]: That makes no sense if Plug has a list of Connections in it. Let's talk online on this one.

Comment (zyga) [inline interfaces/repo.go]: I've fixed everything else. I see why this is controversial, let's talk as suggested and find a better way to do this.

**Comment by niemeyer** [comment]
LGTM

**Comment by niemeyer** [comment]
@chipaca? @pedronis?

File: interfaces/repo.go [hunk 11]
```diff
@@ -449,28 +449,73 @@ func (r *Repository) PlugConnections(snapName, plugName string) []*Slot {
...
+	for plug := range r.slotPlugs[slot] {
+		result = append(result, plug)
+	}
+	sort.Sort(byPlugSnapAndName(result))
+	return result
 }
 
-type bySlotSnapAndName []*Slot
+// Interfaces returns object holding a lists of all the plugs and slots and their connections.
+func (r *Repository) Interfaces() *Interfaces {
```
Comment (mvo5) [inline interfaces/repo.go]: Pardon my ignorance here, the branch is a bit long. Is this covered directly by tests? It seems like it is worthwhile to test if its not already.

File: interfaces/sorting.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+type bySlotRef []SlotRef
```
Comment (mvo5) [inline interfaces/sorting.go]: Is this covered by tests? If not we may consider it - TBH I'm sitting a bit on the fence, its so simple tests look almost like overkill. But then they are probably easy enough (and quick enough) to add :)

**Comment by mvo5** [comment]
This looks good to me, I just have two questions and we need to know what to do about SlotConnections().

File: interfaces/repo.go (hunk 11 above)
Comment (zyga) [inline interfaces/repo.go]: I have test for this, ironically piled up in a separate branch so that size doesn't grow beyond reviewable bounds.

**Comment by zyga** [comment]
@mvo5 SlotConnections and a few other methods are now going away as they are entirely replaced by Interfaces(). I have this piled up after this branch lands.

File: interfaces/sorting.go (hunk 12 above)
Comment (zyga) [inline interfaces/sorting.go]: This is not covered yet. I can conjure a a few quick tests for that.

Comment (zyga) [inline interfaces/sorting.go]: Done

**Pull request merged**